File: apps/desktop/src/lib/control/selectedTaskReviewDecision.ts
import { parseSingleRecordResponse, type QueryFallback } from "./singleRecordResponse";
import type { ControlResponseEnvelopeDto } from "./envelopes";
import type { ControlTaskWorkflowNoEffectsDto } from "./taskWorkflow";
import { invoke } from "@tauri-apps/api/core";

export type SelectedTaskReviewDecisionAction =
  | "accept_evidence"
  | "reject_evidence"
  | "request_changes"
  | "abandon_review";

export type ControlSelectedTaskReviewDecisionNoEffectsDto =
  ControlTaskWorkflowNoEffectsDto & {
    review_mutation_performed: boolean;
  };

export type ControlSelectedTaskReviewDecisionCommandDto = {
  decision_id: string;
  project_id: string;
  task_id: string;
  action: SelectedTaskReviewDecisionAction | string;
  outcome: string;
  expected_revision: string;
  operator_ref: string;
  reviewed_evidence_refs: string[];
  idempotency_key: string;
  reason: string | null;
};

export type ControlSelectedTaskReviewDecisionRefusalDto = {
  kind: string;
  reason: string;
};

export type ControlSelectedTaskReviewDecisionAdmissionDto = {
  admission_id: string;
  decision_id: string;
  project_id: string;
  task_id: string;
  action: SelectedTaskReviewDecisionAction | string;
  status: string;
  command: ControlSelectedTaskReviewDecisionCommandDto | null;
  refusal: ControlSelectedTaskReviewDecisionRefusalDto | null;
  operator_ref: string;
  evidence_refs: string[];
  no_effects: ControlSelectedTaskReviewDecisionNoEffectsDto;
};

export type ControlSelectedTaskReviewDecisionRecordDto = {
  decision_id: string;
  admission_id: string;
  project_id: string;
  task_id: string;
  work_item_refs: string[];
  action: SelectedTaskReviewDecisionAction | string;
  outcome: string;
  operator_ref: string;
  expected_revision: string;
  reviewed_evidence_refs: string[];
  receipt_refs: string[];
  timeline_refs: string[];
  reason_summary: string | null;
  idempotency_key: string;
  status: string;
  blockers: string[];
  duplicate_decision_detected: boolean;
  review_mutation_performed: boolean;
  task_lifecycle_mutation_performed: boolean;
  provider_execution_performed: boolean;
  provider_write_performed: boolean;
  scm_or_forge_mutation_performed: boolean;
  accepted_memory_apply_performed: boolean;
  planning_apply_performed: boolean;
  projection_write_performed: boolean;
  agent_scheduling_performed: boolean;
  ui_effect_performed: boolean;
  raw_provider_material_retained: boolean;
  raw_command_output_retained: boolean;
};

export type SelectedTaskReviewDecisionAdmissionQueryResult =
  | {
      state: "record";
      admission: ControlSelectedTaskReviewDecisionAdmissionDto;
    }
  | QueryFallback;

export type SelectedTaskReviewDecisionApplyQueryResult =
  | {
      state: "record";
      record: ControlSelectedTaskReviewDecisionRecordDto;
    }
  | QueryFallback;

export function readTaskReviewDecisions(
  projectId: string,
  taskId: string,
): Promise<ControlSelectedTaskReviewDecisionRecordDto[]> {
  return invoke<ControlSelectedTaskReviewDecisionRecordDto[]>("read_task_review_decisions", {
    projectId,
    taskId,
  });
}

export function selectedTaskReviewDecisionAdmissionFromResponse(
  response: ControlResponseEnvelopeDto,
): SelectedTaskReviewDecisionAdmissionQueryResult {
  return parseSingleRecordResponse(response, "selected_task_review_decision_admission", "selected task review decision admission", (body) => ({
    state: "record" as const,
    admission: body.admission,
  }));
}

export function selectedTaskReviewDecisionApplyFromResponse(
  response: ControlResponseEnvelopeDto,
): SelectedTaskReviewDecisionApplyQueryResult {
  return parseSingleRecordResponse(response, "selected_task_review_decision_apply", "selected task review decision apply", (body) => ({
    state: "record" as const,
    record: body.record,
  }));
}

File: apps/nucleusd/src/query/typed_response/selected_task_review_decision.rs
use nucleus_server::{
    ControlSelectedTaskReviewDecisionAdmissionDto, ControlSelectedTaskReviewDecisionRecordDto,
};

pub(crate) fn selected_task_review_decision_admission_response_lines(
    label: &str,
    admission: ControlSelectedTaskReviewDecisionAdmissionDto,
) -> Vec<String> {
    let outcome = admission
        .command
        .as_ref()
        .map(|command| command.outcome.as_str())
        .unwrap_or("none");
    let expected_revision = admission
        .command
        .as_ref()
        .map(|command| command.expected_revision.as_str())
        .unwrap_or("none");
    let refusal_kind = admission
        .refusal
        .as_ref()
        .map(|refusal| refusal.kind.as_str())
        .unwrap_or("none");
    let refusal_reason = admission
        .refusal
        .as_ref()
        .map(|refusal| refusal.reason.as_str())
        .unwrap_or("none");

    vec![
        format!("domain={label}"),
        "mode=dry_run".to_owned(),
        format!("admission_id={}", admission.admission_id),
        format!("decision_id={}", admission.decision_id),
        format!("project_id={}", admission.project_id),
        format!("task_id={}", admission.task_id),
        format!("action={}", admission.action),
        format!("status={}", admission.status),
        format!("outcome={outcome}"),
        format!("operator_ref={}", admission.operator_ref),
        format!("expected_revision={expected_revision}"),
        format!("refusal_kind={refusal_kind}"),
        format!("refusal_reason={refusal_reason}"),
        format!("evidence_refs={}", admission.evidence_refs.len()),
        format!(
            "review_mutation_performed={}",
            admission.no_effects.review_mutation_performed
        ),
        format!(
            "task_mutation_performed={}",
            admission.no_effects.task_mutation_performed
        ),
        format!(
            "provider_execution_performed={}",
            admission.no_effects.provider_execution_performed
        ),
        format!(
            "scm_or_forge_mutation_performed={}",
            admission.no_effects.scm_or_forge_mutation_performed
        ),
        "client_can_mutate=false".to_owned(),
    ]
}

pub(crate) fn selected_task_review_decision_apply_response_lines(
    label: &str,
    record: ControlSelectedTaskReviewDecisionRecordDto,
) -> Vec<String> {
    vec![
        format!("domain={label}"),
        "mode=apply".to_owned(),
        format!("decision_id={}", record.decision_id),
        format!("admission_id={}", record.admission_id),
        format!("project_id={}", record.project_id),
        format!("task_id={}", record.task_id),
        format!("action={}", record.action),
        format!("outcome={}", record.outcome),
        format!("status={}", record.status),
        format!("operator_ref={}", record.operator_ref),
        format!("expected_revision={}", record.expected_revision),
        format!("work_item_refs={}", record.work_item_refs.len()),
        format!(
            "reviewed_evidence_refs={}",
            record.reviewed_evidence_refs.len()
        ),
        format!("receipt_refs={}", record.receipt_refs.len()),
        format!("timeline_refs={}", record.timeline_refs.len()),
        format!("blockers={}", record.blockers.len()),
        format!(
            "duplicate_decision_detected={}",
            record.duplicate_decision_detected
        ),
        format!(
            "review_mutation_performed={}",
            record.review_mutation_performed
        ),
        format!(
            "task_lifecycle_mutation_performed={}",
            record.task_lifecycle_mutation_performed
        ),
        format!(
            "provider_execution_performed={}",
            record.provider_execution_performed
        ),
        format!(
            "scm_or_forge_mutation_performed={}",
            record.scm_or_forge_mutation_performed
        ),
        format!(
            "raw_provider_material_retained={}",
            record.raw_provider_material_retained
        ),
        format!(
            "raw_command_output_retained={}",
            record.raw_command_output_retained
        ),
    ]
}
